### ISSUE TITLE
Include successful dependencies only on failed request

### ISSUE THREAD

**Issue by tboby** [microsoft/ApplicationInsights-dotnet#1874]
**Describe your environment.**
* SDK version: 2.12.0
* .NET runtime version (.NET or .NET Core, TargetFramework in the .csproj file): .NET Framework, v4.8
* Hosting Info (IIS/Azure WebApps/etc): ASP.NET MVC, IIS Express
* Platform and OS version: Windows 10 1909

**What are you trying to achieve?**
85%+ of my submitted events and ingested data are SQL dependencies, and I haven't even enabled SQL command logging. While I can disable all non-failing SQL dependencies using an `ITelemetryProcessor`, this means I lose the context of multiple SQL calls during a single failing request.

I think this can be summarised as:
1. A failed dependency should submit
2. A success dependency should submit only if the request as a whole failed (e.g. a later dependency failed)
3. The operation as a whole should follow normal sampling logic.

Documentation explains how to filter at an individual item, and how to influence sampling, however it does not explain how to filter in the context of a request.

I understand this will skew my results, however I cannot afford to pay for 10x data ingestion just to have context for rare events!

**Additional context/attempted solutions**
I can suppress successful dependencies using https://stackoverflow.com/a/41807317/3175307
altered to `dependency?.Type == "SQL" && dependency?.Success == true`.
However, as far as I can tell I can't get the context of whether the parent request succeeded or not.

I experimented with `ITelemetryInitializer`,`SamplingPercentage`, `ExcludeTypes` and `IncludeTypes`, however this is at the wrong scale as I don't want to affect sampling.

The closest solution I found was this article https://blog.ramondeklein.nl/2017/05/05/filtering-application-insights/ where the author manually tracks requests to provide this. However, I can't find any documentation on how the ASP.NET library orders events, so I'm not sure I can safely assume RequestTelemetry will come before DependencyTelemetry.

**Comment by rajkumar-rangaraj** [comment]
@tboby I understand your issue. We don't have a feature inbuilt in application insights SDK to track the complete flow when one of the dependencies fail. Implementing Telemetry processor is a good way to customize for this scenario. 

Normal ASP.NET project, receives request first and then makes backend call. If your application model is based on it, RequestTelemetry gets initiated before DependencyTelemetry and ends after DependencyTelemetry to track the complete request duration.

**Comment by flagbug** [comment]
I think I have the same issue. Basically every time there is a failed request I want to disable sampling for that failed request and include **all** dependency, trace, and exception information for that specific request.

As far as I understand, this doesn't seem to be possible at the moment, correct?

**Comment by cijothomas** [comment]
> I think I have the same issue. Basically every time there is a failed request I want to disable sampling for that failed request and include **all** dependency, trace, and exception information for that specific request.
> 
> As far as I understand, this doesn't seem to be possible at the moment, correct?

Yes its not possible.
The flow is something like this:
Request starts (RequestTelemetry instance created, success/failure = unknown)
make dependency call - (DependencyTelemetry created)
dependency succeeds or fails. (dependencyTelemetry.Sucess is set accordingly)
dependencytelemetry is "Tracked" - sampler is applied to dependencyTelemetry. It doesn't know if the Parent Request is success/failure, so no decision can be made on that. The current sampler makes decision purely based on operationid.
(other telemetry like trace, etc are tracked and has same fate as dependencytelemetry)
Request ends (RequestTelemetry.success/failure is set)
requesttelemetrty is "Tracked" - sampler is applied - at best sampler can decide to retain the RequestTelemetry, but the "child" telemetry's decision was already made and retaining parent RequestTelemetry alone will lead to broken traces.

In short, not possible with any builtin sampler.

**Comment by cijothomas** [comment]
Closing this. Let us know if there are more question.
